### PR TITLE
Fix `gj`/`gk`/`gM` to avoid being affected by non-inline virtual text

### LINKED ISSUE #12028: If 'list' is on, 'below' virtual text which includes 1 or 2 characters are gone
If text is a single character and draw the "below" virtual line.

Then &list is set, all the virtual lines are gone.

If "above", looks like no problem.

And if "below", the problem happens only if text is 1 or 2 characters.

### PR DESCRIPTION
**Pull request by dylanahsmith**
Fixes #12028

## Problem

As mentioned in https://github.com/vim/vim/issues/12028#issuecomment-1966441523, `gj`/`gk` needed to be repeated for each virtual text line to skip over them.  It was actually moving the `w_curswant` position onto the outer virtual text, as if the outer virtual text (and their padding) was a part of the line, even though the actual cursor position would be shifted to stay in the actual text.

I found a similar issue with `gM` including the outer virtual text in the line length used to calculate where the middle of the line is to move the cursor there.

## Solution

I've added a `linetabsize_no_outer` function in order to get the line length with outer virtual text properties excluded.  We can't just exclude text properties entirely, since we do want the influence of inline virtual text in these cases.  I've then used that to get the line length in the functions to handle `gj`/`gk`/`gM` cursor movement.

Note that I had tested `gk` on the first line and `gj` on the last line to ensure the cursor doesn't move, however, I omitted that from the included regression test due to issue #14261 causing the ruler to change when I would otherwise verify that the screen didn't change.  Also, the regression tests screen dumps will need updating if we exclude the 'above' virtual text from the ruler's vcol as I suggested in that issue.

### CODE REVIEW

File: src/charset.c [hunk 1]
```diff
@@ -798,6 +798,40 @@ linetabsize(win_T *wp, linenr_T lnum)
...
+    init_chartabsize_arg(&cts, wp, lnum, 0, line, line);
+
+    int write_idx = 0;
+    for (int read_idx = 0; read_idx < cts.cts_text_prop_count; read_idx++)
+    {
+	textprop_T *tp = &cts.cts_text_props[read_idx];
+	if (tp->tp_col != MAXCOL)
+	{
+	    if (read_idx != write_idx)
+		cts.cts_text_props[write_idx] = *tp;
```
Comment (chrisbra) [inline src/charset.c]: I think you are leaking memory here.

Comment (dylanahsmith) [reply src/charset.c]: Oh, not here, but if `cts.cts_text_prop_count` gets changed to `0` through this filtering, then `cts.cts_text_props` won't be freed by `clear_chartabsize_arg`.

I've pushed a fixup commit to fix the memory leak.  Would you like me to rebase and squash the fixup?

Comment (chrisbra) [reply src/charset.c]: if you could squash, that would be prefered, thanks!

**Comment by chrisbra** [comment]
thanks. I think this makes sense.

**Comment by chrisbra** [comment]
Hi @dylanahsmith Coverity complains about 
grafik

I think this is a false positive, right?

**Comment by dylanahsmith** [comment]
Yeah, that looks like a false positive, since `cts_text_prop_count` is 0 if `cts_text_props` is `NULL`, and has a branch ensuring that if the allocation of `cts_text_props` fails.

**Comment by chrisbra** [comment]
Yes thanks that was also what I thought. Just wanted to get a second opinion about it. I have now marked it false positive in coverity.

**Pull request closed without merging**